routing/payment_state_machine.go

```diff
@@ -0,0 +1,419 @@
+	// found.
+	ErrNotFound = errors.New("payment hash not found")
+)
+
+// paymentStateMachine is the persistent state machine used by the
+// ChannelRouter to track the progress of active payments. It is used to keep
+// information across restarts, needed to properly retry failed paymets, and
+// handle payment errors and successes encountered for a payment sent before
+// the ChannelRouter was restarted.
+type paymentStateMachine struct {
+	db *channeldb.DB
+}
+
+// paymentState is the type used to indicate which state a given payment has in
+// the payment state machine.
+type paymentState byte
+
+const (
+	// paymentStateIntended is the starting state of a payment. A payment
+	// in this state is intended for sending, but haven't yet had a
+	// paymentAttempt sent to the Switch..
+	paymentStateIntended paymentState = 1
+
+	// paymentStateAttemptSent is the state a payment enters after we have
+	// crafted a paymentAttempt for it and sent it to the Switch. In this
+	// state we expect to find the paymentAttempt in the DB, such that we
+	// have the information necessary to handle any results received from
+	// the Switch. After a restart the paymentAttempt can poentially have
+	// been lost by the Switch, in case we can go back to the initial state
+	// and craft a new paymentAttempt to retry.
+	paymentStateAttemptSent paymentState = 2
+
+	// paymentStateSuccess is the state a payment enters after it has
+	// completed successfully. In this state we expect to find the
+	// completed OutgoingPayment in the payments DB. The payment is kept in
+	// this state by the payment state machine to avoid mistakenly
+	// resending a payment to the same payment hash.
+	paymentStateSuccess paymentState = 3
+
+	// paymentStateFailed is the state a payment enters when it has failed
+	// permanently. In this state we will find the error encounterd in the
+	// DB. In this state it is safe to attempt to resend a payment to the
+	// same payment hash.
+	paymentStateFailed paymentState = 4
+
+	// TODO(halseth): timeout/cancel state?
+)
+
+// initPayment registers the given payment hash with the payment state machine.
+// It ensures that the hash is either new, or that the existing payment to this
+// hash is in a state where re-sending it is safe.
+//
+// When this method returns the payment will be in the state
+// paymentStateIntended.
+func (s *paymentStateMachine) initPayment(pHash lntypes.Hash) error {
+	return s.db.Batch(func(tx *bbolt.Tx) error {
+		bucket, err := s.getWriteBucket(tx, pHash)
+		if err != nil {
+			return err
+		}
+
+		// Check if this pHash already has a payment state. We cannot
+		// move it to the intended state if it is already in flight.
+		// Only allow it if this is the first time, or it failed
+		// earlier.
+		v := bucket.Get(paymentStateKey)
+		if v != nil {
+			state := paymentState(v[0])
+			if state != paymentStateFailed {
+				return fmt.Errorf("cannot re-init payment "+
```

Comment:
A concrete error here perhaps?

Reply:
Also a note that if two clients try the same payment, then ideally only the first one is dispatched, while the other blocks waiting on that attempt to complete.

Reply:
> A concrete error here perhaps?

 Fixed here: https://github.com/lightningnetwork/lnd/pull/2761/commits/ba989d0612d5b33ecbd8d7e4b462a580eebf00f1, by mimicking the `ControlTower` errors. This ensure that we don't change the API behavior that much.

> Also a note that if two clients try the same payment, then ideally only the first one is dispatched, while the other blocks waiting on that attempt to complete.

This is also intentionally left out, but can be added later.